actions/actions.py

```diff
@@ -43,6 +44,31 @@
                  }
 
 
+def _check_path_traversal(path_, parent_dir):
+    """Check that 'path_' does not lie outside of the 'parent_dir'.
+
+    This function takes into account possible '../' in 'path_' and also
+    any symlinks that could point somewhere outside the expected 'parent_dir'
+
+    NOTE(mkalcok): This implementation could be improved by using
+                   'os.path.commonpath()'. However it's available only in
+                    py35+.
+
+    :param path_: Path to be tested
+    :param parent_dir: Directory in which the 'path_' must lie
+    :raises: RuntimeError if 'path_' is outside of the 'parent_dir'
+    """
+    full_path = os.path.realpath(path_)
```

Comment:
I've tested the `backup` and `restore` actions by feeding the generated backup as the name of the restore operation, and it failed because of the following symlink:
/var/lib/juju/agents/unit-easyrsa-0/charm/EasyRSA -> /var/lib/juju/agents/unit-easyrsa-0/charm/EasyRSA-3.0.1

The CRITICAL error was:

```
unit-easyrsa-0: 11:12:31 CRITICAL unit.easyrsa/0.juju-log Path traversal detected. '/var/lib/juju/agents/unit-easyrsa-0/charm/EasyRSA-3.0.1/pki/pki/issued/10.5.0.3.crt' tries to travers out of /var/lib/juju/agents/unit-easyrsa-0/charm/EasyRSA/pki/
```

Reply:
strange. I could've sworn that I tested it after the last changes. Nevertheless, this should be easy to fix.